contribute/design/1.x/ndm/20190705-ndm-disk-identification.md

```diff
@@ -0,0 +1,237 @@
+claim possibly leading to data loss
+
+#### Proposed Implementation
+
+User will be able to give a configuration to NDM so that they can choose between the device identification
+method to be used. As an inital step, 3 methods will be provided
+- Ignore the disks which we are unable to uniquely identify.
+- Use the current implementation of disk identification.
+- Use the new implementation of creating GPT parition to identify the disk.
+
+
+##### Disk Identification using GPT partitions
+
+1. We try to generate the UUID using the current implementation. 
+
+2. If the `ID_TYPE` is empty or `ID_MODEL` is `Virtual_disk` or `EphemeralDisk` or `QEMU_HARDDISK` and `ID_WWN` is empty 
+
+    1. Check if the disk has a GPT label available on it.
+    	- If a GPT label is available we will use only that label for UID generation
+    	  of the disk. This will ensure that even if the path is changed, or the disk
+    	  comes up at a different node, the UID generated will be same and the disk
+    	  can be uniquely identified.
+    	- Once the GPT label is available, it is possible to identify partitions
+    	  also on the disk (if it exists). This will work by adding the partition
+    	  number along with the generated UID.
+    	  eg: disk-xxxx-1 (/dev/sdX1), disk-xxxx-2 (/dev/sdX2). 
+    	  This is guaranteed to be unique since the partition numbers won't change
+    	  even if the device path changes
+    2. Check if the disk has a Filesystem label available on it. This is in cases where the 
+       complete disk is formatted with a filesystem without a partition table. If filesystem
+       label is available we use it for generating the UUID.
+    3. If GPT and FS label is not available on the disk, we create a GPT partition table on the disk,
+       and then create a single partition which spans the complete disk. This new partition will
+       be consumed by the users. This will help to identify a virtual disk for movements across
+       the cluster also.
+
+This will be further configurable, so that users can specify how many partition to be created etc. This
+partitions will be automatically by NDM during start-up itself.
+
+Data Cleanup:
+NDM takes care of data clean up, after a BD is released from a BDC. i.e. a complete
+wipe of the disk will be done `wipefs -fa`. Since in case of NDM created GPT labels, 
+only partition is being wiped, it won't cause the labels to be removed.
+
+##### Workflow
+```
++------------+
+|check config|
+|    if GPT  |
+|  method is |
+|  selected  |
+|            |
++------------+
+     |yes
+     |
++----v-----+
+|UID = WWN +|
+|    Model +|
+|   Serial +|
+|   Vendor  |
+|           |
++-----------+
+     |
+     |
++----v-------------+        +-----------+      +-----------+      +-------------------+
+|    If            |        |  Check    |      |  Check    |      | Create GPT        |
+|ID_TYPE is empty  |   Yes  | GPT label |  No  | FS label  |  No  | partition         |
```

Comment:
Shared, dual-port NVMe devices might not have an ID_TYPE. They do have a WWN, serial, vendor, and model. We can infer they are NVMe by matching the DEVNAME, but I see no other udev environment variable that will identify them as being NVMe.

The current code sees lack of ID_TYPE and adds the hostname into the UID. This is a problem because the drive appears once per host. Is it sufficient to use a valid, non-known-to-be-ephemeral WWN and SERIAL?

Reply:
In the new algorithm, we check for WWN and consider it to be unique. So shouldn't the problem with dual Port NVMe be mitigated?

Reply:
the new algorithm sounds good